src/fundamental_interface.jl

```diff
@@ -0,0 +1,275 @@
+###############################################################################
+#
+#   basic_interface.jl : basic interface for AbstractAlgebra
+#
+###############################################################################
+
+# TODO: Move more generic functions to this file.
+
+###############################################################################
+#
+#   Parents, elements and data type methods
+#
+###############################################################################
+
+@doc Markdown.doc"""
+    parent(a)
+
+Return parent object of given element $a$.
+
+# Examples
+```jldoctest; setup = :(using AbstractAlgebra)
+julia> G = SymmetricGroup(5); g = Perm([3,4,5,2,1])
+(1,3,5)(2,4)
+
+julia> parent(g) == G
+true
+
+julia> S, x = LaurentSeriesRing(ZZ, 3, "x")
+(Laurent series ring in x over Integers, x + O(x^4))
+
+julia> parent(x) == S
+true
+```
+"""
+function parent end
+
+# TODO: Give example
+@doc Markdown.doc"""
+    elem_type(parent)
+    elem_type(parent_type)
+
+Given a parent object (or its type), return the type of its elements.
+
+# Example
+```jldoctest; setup = :(using AbstractAlgebra)
+julia> S, x = PowerSeriesRing(QQ, 2, "x")
+(Univariate power series ring in x over Rationals, x + O(x^3))
+
+julia> elem_type(S) == typeof(x)
+true
+```
+"""
+elem_type(x)  = elem_type(typeof(x))
+elem_type(T::DataType) = throw(MethodError(elem_type, (T,)))
+
+@doc Markdown.doc"""
+    parent_type(element)
+    parent_type(element_type)
+
+Given an element (or its type), return the type of its parent object.
+
+# Example
+```jldoctest; setup = :(using AbstractAlgebra)
+julia> R, x = PolynomialRing(ZZ, "x")
+(Univariate Polynomial Ring in x over Integers, x)
+
+julia> S = MatrixSpace(R, 2, 2)
+Matrix Space of 2 rows and 2 columns over Univariate Polynomial Ring in x over Integers
+
+julia> a = rand(S, 0:1, 0:1);
+
+julia> parent_type(a) == typeof(S)
+true
+```
+"""
+parent_type(x) = parent_type(typeof(x))
+parent_type(T::DataType) = throw(MethodError(parent_type, (T,)))
+
+@doc Markdown.doc"""
+    base_ring(a::T)
+
+Return base ring $R$ of given element or parent $a$.
+
+# Examples
+```jldoctest; setup = :(using AbstractAlgebra)
+julia> S, x = PolynomialRing(QQ, "x")
+(Univariate Polynomial Ring in x over Rationals, x)
+
+julia> base_ring(S) == QQ
+true
+
+julia> R = GF(7)
+Finite field F_7
+
+julia> base_ring(R)
+Union{}
+```
+"""
+function base_ring end
+
+###############################################################################
+#
+#   Special elements
+#
+###############################################################################
+
+@doc Markdown.doc"""
+    one(a::T)
+
+Return the (multiplicative) identity in the family of $a$.
```

Comment:
Don't you mean "parent of $a$"? Otherwise, what is the "family" of $a$? Where is this notion defined?

Reply:
So this was a word I coined. I understand your confusion, since it is not written anywhere else and not explained.

The word family refers to the union/collection of parent and element. This is because you can call the function like
```
julia> one(QQ(1, 3)) == one(QQ)
true
```

Reply:
OK fair enough. If we need to talk a lot about these things, then I guess it'd make sense to define a term; but as long as it's just a few places, I am a bit wary about this, as a user must be aware of it (but I guess one could say something liek "... in the family of $a$ (see XYZ for definition of family)" or so.

The alternative might be to split this into two docstrings, or at least two signatures.... perhaps vaguely like this:
```
    one(a::elmT)

Return the multiplicative identify of the the parent of the element $a$.
```
```
    one(a::parentT)

Return the multiplicative identify of the algebraic structure $a$.
```

Reply:
Can one just say `Return the multiplicative identity in the algebraic structure as $a$.`? Maybe this is wrong language use, but it sounds good to me in order to include both algebraic structures and elements.

Reply:
I guess there should be two docstrings, one for `RingElem` and one for `Ring`.

Reply:
I think I have solved this in a nice way. I will close this, but please unresolve if this solution is not desirable or not good enough.